src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Binding.BindInfo;
+using Microsoft.PowerFx.Core.Entities;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Types.Enums;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx
+{
+    /// <summary>
+    /// Used between Symbol Table resolver and IR. Enables describing constants or 
+    /// globals that are the same across evals. 
+    /// Object provided by resolver data in <see cref="NameLookupInfo.Data"/>.
+    /// IR then recognizes this and will look these values and will retrieve the Value at runtime. 
+    /// </summary>
+    internal interface ICanGetValue
+    {
+        FormulaValue Value { get; }
+    }
+
+    /// <summary>
+    /// Used between Symbol Table resolver and IR.
+    /// NameInfo data to associate a symbol at bind time with a runtime config at runtime. 
+    /// Object provided by resolver data in <see cref="NameLookupInfo.Data"/>.
+    /// IR then recognizes this and will look these values up via ReadOnlySymbolValues.TryGetValue.
+    /// </summary>
+    internal class NameSymbol
+    {
+        public NameSymbol(string name)
+        {
+            Name = name;
+        }
+        
+        public string Name { get; private set; }
+    }
+
+    /// <summary>
+    /// Provides symbols to the engine. This includes variables (locals, globals), enums, options sets, and functions.
+    /// SymbolTables are mutable to support sessionful scenarios and can abe chained together. 
+    /// This is a publicly facing class around a <see cref="INameResolver"/>.
+    /// </summary>
+    [DebuggerDisplay("{DebugName}")]
+    public class SymbolTable : ReadOnlySymbolTable
+    {
+        // Expose public setters
+        public new ReadOnlySymbolTable Parent
+        {
+            get => _parent;
+            set
```

Comment:
>set [](http://example.com/codeflow?start=12&length=3)

Consider: `init` so it's impossible to define loops, and this needs to be set either by the constructor or object initializer

Reply:
That's a good idea - good way to get compile-time support and we can always loosen it later. 

we have a test for loop detection, but compile support is better.